File: DataRanker_Backend/api/routes/pipeline.py
# api/routes/pipeline.py — /run-pipeline endpoint

from fastapi import APIRouter, File, Form, HTTPException, UploadFile
from fastapi.responses import Response
from flask import json

from core.config import COLUMN_MAPPING
from services.formatter import run_format
from services.mapper import run_mapper
from services.ranker import run_ranking
from utils.file_utils import read_bytes, save_upload, temp_workspace, workspace_path

router = APIRouter()

XLSX_MEDIA_TYPE = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"

@router.get("/column-mapping")
async def get_column_mapping():
    return COLUMN_MAPPING

@router.post("/run-pipeline")
async def run_pipeline(
    query_results: UploadFile = Form(...),
    industry_mapping: UploadFile = File(...),
    kpi_library: UploadFile = File(...),
    mapping_json: str = File(...),
) -> Response:
    
    list_of_mappings = json.loads(mapping_json)
    dynamic_mapping = {k: v for d in list_of_mappings for k, v in d.items()}

    with temp_workspace() as tmpdir:
        try:
            # ── Save uploads ──────────────────────────────────────────────────
            query_path   = workspace_path(tmpdir, "query-results.csv")
            mapping_path = workspace_path(tmpdir, "industry-mapping.xlsx")
            kpi_path     = workspace_path(tmpdir, "KPI_Library.xlsx")

            save_upload(query_results,    query_path)
            save_upload(industry_mapping, mapping_path)
            save_upload(kpi_library,      kpi_path)

            # ── Stage 1: Format ───────────────────────────────────────────────
            formatted_path = workspace_path(tmpdir, "Formatted_Data.xlsx")
            run_format(query_path, formatted_path, dynamic_mapping)

            # ── Stage 2: Map industries ───────────────────────────────────────
            mapped_path = workspace_path(tmpdir, "Mapped_query-results.csv")
            run_mapper(formatted_path, mapping_path, mapped_path)

            # ── Stage 3: Rank ─────────────────────────────────────────────────
            output_path = workspace_path(tmpdir, "Final_Ranked_Report.xlsx")
            run_ranking(mapped_path, kpi_path, output_path)

            # Read bytes NOW, while the temp dir still exists
            file_bytes = read_bytes(output_path)

        except Exception as exc:
            raise HTTPException(status_code=500, detail=str(exc)) from exc

    # temp dir is deleted here — safe because we already have the bytes
    return Response(
        content=file_bytes,
        media_type=XLSX_MEDIA_TYPE,
        headers={"Content-Disposition": 'attachment; filename="Final_Ranked_Report.xlsx"'},
    )

File: server/routes/pipeline.js
// routes/pipeline.js — GET /column-mapping and POST /run-pipeline.

const express = require("express");
const multer = require("multer");

const { COLUMN_MAPPING } = require("../core/config");
const { runFormat } = require("../services/formatter");
const { runMapper } = require("../services/mapper");
const { runRanking } = require("../services/ranker");
const { getOrSeedLibrary, toRankerRows } = require("../services/kpiLibrary");
const { getIndustryMappingBuffer } = require("../services/industryMapping");
const { requireAuth } = require("../middleware/auth");

const router = express.Router();

const XLSX_MEDIA_TYPE =
  "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet";

const MAX_UPLOAD_BYTES = 15 * 1024 * 1024; // 15 MB per file

const upload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: MAX_UPLOAD_BYTES, files: 1 },
});

// Only the query export is uploaded now; the industry-mapping workbook is bundled
// with the backend (services/industryMapping.js).
const uploadFields = upload.fields([{ name: "query_results", maxCount: 1 }]);

// Wrap the multer middleware so a MulterError (e.g. file too large) becomes a
// friendly JSON response instead of an unhandled 500.
function handleUpload(req, res, next) {
  uploadFields(req, res, (err) => {
    if (err) {
      if (err instanceof multer.MulterError) {
        const status = err.code === "LIMIT_FILE_SIZE" ? 413 : 400;
        return res
          .status(status)
          .json({ detail: "Uploaded file is too large (max 15 MB)." });
      }
      return res.status(400).json({ detail: "File upload failed." });
    }
    return next();
  });
}

router.get("/column-mapping", (req, res) => {
  res.json(COLUMN_MAPPING);
});

router.post(
  "/run-pipeline",
  requireAuth, // free tier = any signed-in user
  handleUpload,
  async (req, res) => {
    try {
      const files = req.files || {};
      const queryFile = files.query_results && files.query_results[0];

      if (!queryFile) {
        throw new Error("Missing the required query results file");
      }

      // KPI library now comes from the signed-in user's saved set (seeded with
      // defaults on first use), replacing the uploaded KPI Excel sheet.
      const lib = await getOrSeedLibrary(req.user._id);
      const kpiRows = toRankerRows(lib.rows);

      // Parse mapping_json exactly like the Python: JSON array of single-key
      // objects merged into one dynamic_mapping object.
      const listOfMappings = JSON.parse(req.body.mapping_json);
      const dynamicMapping = listOfMappings.reduce(
        (acc, d) => Object.assign(acc, d),
        {}
      );

      // Stage 1 → Stage 2 → Stage 3 (in-memory). The industry-mapping workbook
      // is the backend's bundled copy, not an upload.
      const formatted = runFormat(queryFile.buffer, dynamicMapping);
      const mapped = runMapper(formatted, getIndustryMappingBuffer());
      const fileBytes = await runRanking(mapped, kpiRows);

      res.status(200);
      res.setHeader("Content-Type", XLSX_MEDIA_TYPE);
      res.setHeader(
        "Content-Disposition",
        'attachment; filename="Final_Ranked_Report.xlsx"'
      );
      res.send(fileBytes);
    } catch (exc) {
      res.status(500).json({ detail: String(exc && exc.message ? exc.message : exc) });
    }
  }
);

module.exports = router;
